homeassistant/components/homewizard/sensor.py

```diff
@@ -426,17 +421,106 @@ class HomeWizardSensorEntityDescription(
+        key="Warm water meter",
+        translation_key="total_liter_m3",
+        suggested_device_class=SensorDeviceClass.WATER,
+        state_class=SensorStateClass.TOTAL_INCREASING,
+    ),
+    ExternalDevice.DeviceType.WATER_METER: HomeWizardExternalSensorEntityDescription(
+        key="Water meter",
+        translation_key="total_liter_m3",
+        suggested_device_class=SensorDeviceClass.WATER,
+        state_class=SensorStateClass.TOTAL_INCREASING,
+    ),
+    ExternalDevice.DeviceType.INLET_HEAT_METER: HomeWizardExternalSensorEntityDescription(
+        key="Inlet heat meter",
+        translation_key="total_energy_gj",
+        suggested_device_class=SensorDeviceClass.ENERGY,
+        state_class=SensorStateClass.TOTAL_INCREASING,
+    ),
+}
+
+
 async def async_setup_entry(
     hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
 ) -> None:
     """Initialize sensors."""
     coordinator: HWEnergyDeviceUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
 
-    async_add_entities(
+    # Migrate original gas meter sensor to ExternalDevice
+    ent_reg = er.async_get(hass)
+
+    if (
+        entity_id := ent_reg.async_get_entity_id(
+            Platform.SENSOR, DOMAIN, f"{entry.unique_id}_total_gas_m3"
+        )
+    ) and coordinator.data.data.gas_unique_id is not None:
+        ent_reg.async_update_entity(
+            entity_id,
+            new_unique_id=f"{DOMAIN}_{coordinator.data.data.gas_unique_id}",
+        )
+
+    # Migrate original gas meter identifier sensor to ExternalDevice
+    if (
+        entity_id := ent_reg.async_get_entity_id(
+            Platform.SENSOR, DOMAIN, f"{entry.unique_id}_gas_unique_id"
+        )
+    ) and coordinator.data.data.gas_unique_id is not None:
+        ent_reg.async_update_entity(
+            entity_id,
+            new_unique_id=f"{DOMAIN}_{coordinator.data.data.gas_unique_id}_meter_identifier",
+        )
+
+    # Initialize default sensors
+    entities: list = [
         HomeWizardSensorEntity(coordinator, entry, description)
         for description in SENSORS
         if description.has_fn(coordinator.data.data)
-    )
+    ]
+
+    # Initialize external devices
+    if coordinator.data.data.external_devices is not None:
+        for unique_id, device in coordinator.data.data.external_devices.items():
+            if description := EXTERNAL_SENSORS.get(device.meter_type):
+                device_info = DeviceInfo(
+                    identifiers={(DOMAIN, unique_id)},
+                    name=description.key,
+                    manufacturer="HomeWizard",
+                    model=coordinator.data.device.product_type,
+                )
+                if coordinator.data.device.serial is not None:
+                    device_info[ATTR_VIA_DEVICE] = (
+                        DOMAIN,
+                        coordinator.data.device.serial,
+                    )
```

Comment:
This is an copy what is found in entity.py, but different. Not sure how and if we can improve that.

https://github.com/home-assistant/core/blob/cab30085c5987cb4cd21bc4e8135a52de86df064/homeassistant/components/homewizard/entity.py#L20-L30